server/auth/range_perm_cache.go

```diff
@@ -103,34 +103,50 @@ func checkKeyPoint(lg *zap.Logger, cachedPerms *unifiedRangePermissions, key []b
 	return false
 }
 
-func (as *authStore) isRangeOpPermitted(tx AuthReadTx, userName string, key, rangeEnd []byte, permtyp authpb.Permission_Type) bool {
+func (as *authStore) isRangeOpPermitted(userName string, key, rangeEnd []byte, permtyp authpb.Permission_Type) bool {
 	// assumption: tx is Lock()ed
-	_, ok := as.rangePermCache[userName]
+	as.rangePermCacheMu.RLock()
+	defer as.rangePermCacheMu.RUnlock()
+
+	rangePerm, ok := as.rangePermCache[userName]
 	if !ok {
-		perms := getMergedPerms(tx, userName)
-		if perms == nil {
-			as.lg.Error(
-				"failed to create a merged permission",
-				zap.String("user-name", userName),
-			)
-			return false
-		}
-		as.rangePermCache[userName] = perms
+		as.lg.Error(
+			"user doesn't exist",
+			zap.String("user-name", userName),
+		)
+		return false
 	}
 
 	if len(rangeEnd) == 0 {
-		return checkKeyPoint(as.lg, as.rangePermCache[userName], key, permtyp)
+		return checkKeyPoint(as.lg, rangePerm, key, permtyp)
 	}
 
-	return checkKeyInterval(as.lg, as.rangePermCache[userName], key, rangeEnd, permtyp)
+	return checkKeyInterval(as.lg, rangePerm, key, rangeEnd, permtyp)
 }
 
-func (as *authStore) clearCachedPerm() {
+func (as *authStore) refreshRangePermCache() {
+	as.rangePermCacheMu.Lock()
+	defer as.rangePermCacheMu.Unlock()
+
+	tx := as.be.ReadTx()
+	tx.Lock()
+	defer tx.Unlock()
+
 	as.rangePermCache = make(map[string]*unifiedRangePermissions)
-}
 
-func (as *authStore) invalidateCachedPerm(userName string) {
-	delete(as.rangePermCache, userName)
+	users := as.be.GetAllUsers()
```

Comment:
Aren't we afraid that this path might be costly ? 
I don't know how many user's we have in practice... but if we have series of permission edits (e.g. 
N role Adds, we will have `O(N^2)` actions to be performed on the critical apply code-path. 

What's the reasoning for proactive cache refresh vs. invalidation and `lazy` cache population ?

Reply:
> What's the reasoning for proactive cache refresh vs. invalidation and lazy cache population ?

I selected proactive cache refresh just for simplicity. I know an etcd cluster which initializes permission configuration just after initialization and doesn't udpate it often so won't be so problematic even the operation becomes costly. But as you say having the costly operation in the apply path is potentially dangerous. I'll check performance of the method.

Reply:
I've been fixing a functional issue of this PR, not evaluating performance. I think I'll be able to evaluate performance tomorrow...

Reply:
I'm comparing the main branch and this PR's branch with 100 users and 10 roles each on my local laptop (prepared by this script: https://gist.github.com/mitake/3438b6b7fb64d3f07f3d4282951645e0).
After preparing the dummy users and roles, just creating one user by `etcdctl user add <username>`, which involves `as.refreshRangePermCache()`, didn't show significant difference on each branch.

I measured wall clock time for the function with a change like below and it was about 200 us for each call:
`22:33:43 etcd1 | {"level":"info","ts":"2022-05-11T22:33:43.811+0900","caller":"auth/range_perm_cache.go:153","msg":"elapsed time for refreshRangePermCache: ","elapsed":"195.75µs"}`

```
diff --git a/server/auth/range_perm_cache.go b/server/auth/range_perm_cache.go
index 35fcfe120..43a85b12a 100644
--- a/server/auth/range_perm_cache.go
+++ b/server/auth/range_perm_cache.go
@@ -18,6 +18,7 @@ import (
        "go.etcd.io/etcd/api/v3/authpb"
        "go.etcd.io/etcd/pkg/v3/adt"
        "go.uber.org/zap"
+       "time"
 )

 func getMergedPerms(tx AuthReadTx, userName string) *unifiedRangePermissions {
@@ -125,6 +126,7 @@ func (as *authStore) isRangeOpPermitted(userName string, key, rangeEnd []byte, p
 }

 func (as *authStore) refreshRangePermCache(tx AuthReadTx) {
+       start := time.Now()
        // Note that every authentication configuration update calls this method and it invalidates the entire
        // rangePermCache and reconstruct it based on information of users and roles stored in the backend.
        // This can be a costly operation. TODO(mitake): check how it is costly actually.
@@ -146,6 +148,10 @@ func (as *authStore) refreshRangePermCache(tx AuthReadTx) {
                }
                as.rangePermCache[userName] = perms
        }
+
+       elapsed := time.Since(start)
+       as.lg.Info("elapsed time for refreshRangePermCache: ",
+               zap.Duration("elapsed", elapsed))
 }

 type unifiedRangePermissions struct {
```

I'll evaluate with different parameters but I guess the function isn't adding significant cost.

Reply:
I did a same test with 1000 users and 10 roles for each. The cost for refreshing perm range cache was about 5ms - 10ms: https://gist.github.com/mitake/135b9b51bf05dcd2fb020e17bded5fe8
I think it's safe to conclude the overhead is acceptable. How do you think @ptabor ?